### PR TITLE
Support pluginManagement block in precompiled Groovy settings plugins

### PR DESCRIPTION
**Pull request by rieske**
https://github.com/gradle/gradle/issues/15416

This change addresses two things:
- `pluginManagement` block support in precompiled Groovy settings plugins
- forbid `buildscript` block at compile time in all precompiled Groovy plugins - for precompiled plugins classpath is configured via the project dependencies. This is how Kotlin DSL precompiled plugins work as well (except that there failures happen at runtime (this should probably be improved, but that's a separate topic)).

Precompiled plugin compilation (applies to buildscripts as well) happens in two compilation passes:
1st pass extracts only `buildscript`, `pluginManagement` and `plugins` blocks from the buildscript and compiles those.
2nd pass compiles the body of the buildscript which is everything except the above blocks.

Current implementation assumed that only `plugins {}` block is compiled in the first pass of precompiled Groovy plugins compilation.
As such, the result from the first-pass compilation was used only to extract the plugin ids to be applied in the resulting precompiled plugin.
With this change, the result of the first compilation pass is made part of the resulting precompiled plugin and is executed in the generated Plugin adapter class.

Because the result of the first pass can contain `buildscript` block as well (which does not make sense in context of precompiled script plugins), which was silently ignored
in current implementation, the change also addresses failing early if such block is used in a precompiled plugin.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/SettingsInternal.java [hunk 1]
```diff
@@ -67,6 +68,8 @@
 
     File getBuildSrcDir();
 
+    ServiceRegistry getServices();
```
Comment (rieske) [inline subprojects/core/src/main/java/org/gradle/api/internal/SettingsInternal.java:71]: Precompiled Groovy settings plugins were using the wrong services scope - GradleScopeServices - which should be used by init scripts. Exposing the settings service registry here allows precompiled settings plugins to fetch the right one.

File: subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/plugins/PrecompiledGroovyPluginsIntegrationTest.groovy [hunk 2]
```diff
@@ -371,9 +367,41 @@ class PrecompiledGroovyPluginsIntegrationTest extends AbstractIntegrationSpec {
...
+            buildscript {
+                dependencies {}
+            }
+            println("foo settings plugin applied")
+        """
+
+        then:
+        fails('help')
+        failureCauseContains("The `buildscript` block is not supported in Groovy script plugins, please use the `plugins` block or project level dependencies.")
+    }
```
Comment (rieske) [inline subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/plugins/PrecompiledGroovyPluginsIntegrationTest.groovy:403]: Those two make sure the compilation fails with a message that is similar to one produced by the Kotlin counterpart.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/FirstPassPrecompiledScript.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import groovy.lang.Closure;
+import org.gradle.plugin.use.internal.PluginsAwareScript;
+
+public abstract class FirstPassPrecompiledScript extends PluginsAwareScript {
+
+    @SuppressWarnings("rawtypes")
+    @Override
+    public void buildscript(Closure configureClosure) {
+        throw new IllegalStateException("The `buildscript` block is not supported in Groovy script plugins, please use the `plugins` block or project level dependencies.");
```
Comment (donat) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/FirstPassPrecompiledScript.java:27]: Maybe.
```suggestion
        throw new IllegalStateException("The `buildscript` block is not supported in Groovy script plugins. Use the `plugins` block or project level dependencies instead.");
```

[the remark this answers is not in the document]
Comment (donat) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/FirstPassPrecompiledScript.java:27]: Same with the other exceptions.

**Review by donat — APPROVED** [review]
The PR is good to go in terms of functionality imho.

At the same time, I'm missing the documentation on this feature, and on precompiled settings script plugins as well. We should revisit the documentation and explain how this feature should be used.

There are also aspects of this feature that I don't fully understand:
- why do I need to manually include the buildSrc jar in the settings script classpath? I usually don't have to do that with precompiled script plugins.
- why can't I use the `plugins {}` block to apply the plugin? I can use that with precompiled script plugins.
- (a consequence of the previous two items) why shouldn't a plugin developer prefer a traditional settings plugin, maybe declared as an included build?

It's probably my lack of knowledge and the difficulty of this domain, but I still need to ask.

**Comment by rieske** [comment]
@donat 

We do miss documentation on differences in Project versus Settings plugins in general - that's true and we should fix it.

To answer the questions raised:
> why do I need to manually include the buildSrc jar in the settings script classpath? I usually don't have to do that with precompiled script plugins.

That's mostly historic - included builds could not contribute settings plugins to the project until recently (7.0 actually): https://docs.gradle.org/7.0-milestone-3/release-notes.html#plugin-development-improvements
Thus, original tests for the precompiled setting plugins used "a workaround" to package the settings plugin in a jar using a separate project and then include that jar in the settings script of the consuming project using the `buildscript {}` block.
Now, it should be possible to rewrite those tests to use an included build instead - I'll give it a shot.

> why can't I use the plugins {} block to apply the plugin? I can use that with precompiled script plugins.

That's because the jar containing the plugin is included using the `buildscript {}` block - same case for project plugins when using the ["legacy" plugin application](https://docs.gradle.org/current/userguide/plugins.html#sec:old_plugin_application).
With https://docs.gradle.org/7.0-milestone-3/release-notes.html#plugin-development-improvements this should no longer be needed and `plugins {}` block should be used. As above, if I can make the tests use included builds, this will automatically go away.

> (a consequence of the previous two items) why shouldn't a plugin developer prefer a traditional settings plugin, maybe declared as an included build?

Like above, with changes in 7.0: https://docs.gradle.org/7.0-milestone-3/release-notes.html#plugin-development-improvements the plugin developer will be able to use an included build to include a settings plugin (regardless if it is "traditional" or precompiled script plugin).

**Comment by jjohannes** [comment]
Great to see this fix!

I noticed a difference between Kotlin and Groovy precompiled script plugins. But I am not sure if...
- this is a difference we can/should just accept
- the Kotlin implementation is "tool lenient"
- it should be adjusted as part of this PR

In Kotlin, I can use a method defined in the build script in the plugin management block (if this is good practice, I don't know...):

 ```
pluginManagement {
    foo() // compiles
}

fun foo() {}
```

In Groovy this is not possible. The method is not found

```
pluginManagement {
    foo() // fails to compile
}

def foo() {}
```

Here is where it is dogfooded in our own build (#15526). Generally, both solutions are ugly. (How the scenario can be solved better is a different question.) 

Kotlin:  https://github.com/gradle/gradle/blob/jjohannes/idiomatic/gradlebuild/split/build-logic/build-logic-base/settings-plugin/src/main/kotlin/gradlebuild.repositories.settings.gradle.kts#L22-L44
Groovy (with this PR):
https://github.com/gradle/gradle/blob/jjohannes/idiomatic/gradlebuild/split-fix-16625/build-logic/build-logic-base/settings-plugin/src/main/groovy/gradlebuild.settings-plugins.settings.gradle#L21-L40

**Comment by rieske** [comment]
Given that
```
pluginManagement {
    foo()
}
def foo() {}
```
does not compile either with Kotlin or Groovy settings scripts (not settings plugins), I'd lean towards saying that Kotlin plugins are too lenient in this case.

**Comment by jjohannes** [comment]
> does not compile either with Kotlin or Groovy settings scripts (not settings plugins), I'd lean towards saying that Kotlin plugins are too lenient in this case.

Thanks for checking this. I agree.

**Review by jjohannes — APPROVED** [review]

**Comment by rieske** [comment]
@bot-gradle test and merge this

**Pull request merged**